### PR TITLE
Fix segfault in oversized objects check in combination with var.

### LINKED ISSUE #9404: Segfault in solidity::langutil::SourceLocation
## Description
Consider the following POC:
```solidity
contract b {
    struct c {
        uint [ 2 ** 253 ] a ;
    }
    c d ;
    function e () {
        var d = d ;
    }
}
```

Run with `solc --bin -o poc poc.sol` with an solc built with asan and we got 
```
ASAN:DEADLYSIGNAL
=================================================================
==141944==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000018 (pc 0x56511f89dc5f bp 0x7ffdf4954c50 sp 0x7ffdf4954c40 T0)
==141944==The signal is caused by a READ memory access.
==141944==Hint: address points to the zero page.
    #0 0x56511f89dc5e in solidity::langutil::SourceLocation::SourceLocation(solidity::langutil::SourceLocation const&) (/home/yongheng/solidity/asan/solc/solc+0x4f3c5e)
    #1 0x56511fcf4e93 in boost::error_info<solidity::langutil::tag_sourceLocation, solidity::langutil::SourceLocation>::error_info(solidity::langutil::SourceLocation const&) (/home/yongheng/solidity/asan/solc/solc+0x94ae93)
    #2 0x5651207c3185 in solidity::langutil::ErrorReporter::error(solidity::langutil::ErrorId, solidity::langutil::Error::Type, solidity::langutil::SourceLocation const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/home/yongheng/solidity/asan/solc/solc+0x1419185)
    #3 0x5651207c2fb6 in solidity::langutil::ErrorReporter::warning(solidity::langutil::ErrorId, solidity::langutil::SourceLocation const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/home/yongheng/solidity/asan/solc/solc+0x1418fb6)
    #4 0x56511fb89b0c in solidity::frontend::TypeChecker::visit(solidity::frontend::VariableDeclaration const&) (/home/yongheng/solidity/asan/solc/solc+0x7dfb0c)
    #5 0x56511fc033de in solidity::frontend::VariableDeclaration::accept(solidity::frontend::ASTConstVisitor&) const (/home/yongheng/solidity/asan/solc/solc+0x8593de)
    #6 0x56511fb9e30d in solidity::frontend::TypeChecker::visit(solidity::frontend::VariableDeclarationStatement const&) (/home/yongheng/solidity/asan/solc/solc+0x7f430d)
    #7 0x56511fc0797e in solidity::frontend::VariableDeclarationStatement::accept(solidity::frontend::ASTConstVisitor&) const (/home/yongheng/solidity/asan/solc/solc+0x85d97e)
    #8 0x56511fc1d0f4 in void solidity::frontend::ASTNode::listAccept<std::shared_ptr<solidity::frontend::Statement> >(std::vector<std::shared_ptr<solidity::frontend::Statement>, std::allocator<std::shared_ptr<solidity::frontend::Statement> > > const&, solidity::frontend::ASTConstVisitor&) (/home/yongheng/solidity/asan/solc/solc+0x8730f4)
    #9 0x56511fc0525f in solidity::frontend::Block::accept(solidity::frontend::ASTConstVisitor&) const (/home/yongheng/solidity/asan/solc/solc+0x85b25f)
    #10 0x56511fb859f7 in solidity::frontend::TypeChecker::visit(solidity::frontend::FunctionDefinition const&) (/home/yongheng/solidity/asan/solc/solc+0x7db9f7)
    #11 0x56511fc02e3a in solidity::frontend::FunctionDefinition::accept(solidity::frontend::ASTConstVisitor&) const (/home/yongheng/solidity/asan/solc/solc+0x858e3a)
    #12 0x56511fb7bca1 in solidity::frontend::TypeChecker::visit(solidity::frontend::ContractDefinition const&) (/home/yongheng/solidity/asan/solc/solc+0x7d1ca1)
    #13 0x56511fc019da in solidity::frontend::ContractDefinition::accept(solidity::frontend::ASTConstVisitor&) const (/home/yongheng/solidity/asan/solc/solc+0x8579da)
    #14 0x56511fc1b71a in void solidity::frontend::ASTNode::listAccept<std::shared_ptr<solidity::frontend::ASTNode> >(std::vector<std::shared_ptr<solidity::frontend::ASTNode>, std::allocator<std::shared_ptr<solidity::frontend::ASTNode> > > const&, solidity::frontend::ASTConstVisitor&) (/home/yongheng/solidity/asan/solc/solc+0x87171a)
    #15 0x56511fc01283 in solidity::frontend::SourceUnit::accept(solidity::frontend::ASTConstVisitor&) const (/home/yongheng/solidity/asan/solc/solc+0x857283)
    #16 0x56511fb7ad63 in solidity::frontend::TypeChecker::checkTypeRequirements(solidity::frontend::SourceUnit const&) (/home/yongheng/solidity/asan/solc/solc+0x7d0d63)
    #17 0x56511f8138f9 in solidity::frontend::CompilerStack::analyze() (/home/yongheng/solidity/asan/solc/solc+0x4698f9)
    #18 0x56511f815b70 in solidity::frontend::CompilerStack::parseAndAnalyze() (/home/yongheng/solidity/asan/solc/solc+0x46bb70)
    #19 0x56511f8168c4 in solidity::frontend::CompilerStack::compile() (/home/yongheng/solidity/asan/solc/solc+0x46c8c4)
    #20 0x56511f5bd843 in solidity::frontend::CommandLineInterface::processInput() (/home/yongheng/solidity/asan/solc/solc+0x213843)
    #21 0x56511f66200e in main (/home/yongheng/solidity/asan/solc/solc+0x2b800e)
    #22 0x7fb5165b7b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
    #23 0x56511f5a5e09 in _start (/home/yongheng/solidity/asan/solc/solc+0x1fbe09)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/home/yongheng/solidity/asan/solc/solc+0x4f3c5e) in solidity::langutil::SourceLocation::SourceLocation(solidity::langutil::SourceLocation const&)
==141944==ABORTING
```
## Environment

- Compiler version:Version: 0.6.12-develop.2020.7.13+commit.0212a16b.Linux.g++
- Target EVM version (as per compiler settings):
- Framework/IDE (e.g. Truffle or Remix):
- EVM execution environment / backend / blockchain client:
- Operating system: Ubuntu 16

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes https://github.com/ethereum/solidity/issues/9404

### CODE REVIEW

**Comment by ekpyron** [comment]
Really interesting interaction going on here :-).

File: test/libsolidity/syntaxTests/iceRegressionTests/oversized_var.sol [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+        uint [ 2 ** 253 ] a ;
+    }
+
+    c d ;
+    function e () {
+        var d = d ;
+    }
+}
+// ----
+// SyntaxError 4937: (79-120): No visibility specified. Did you intend to add "public"?
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/iceRegressionTests/oversized_var.sol]: Can you simplify the test a little to only contain the relevant errors / warnings?

Comment (ekpyron) [reply test/libsolidity/syntaxTests/iceRegressionTests/oversized_var.sol]: Yeah - ICE regression tests are always a bit weird, so I usually don't bother too much, but this one can really go, that's true.

Comment (ekpyron) [reply test/libsolidity/syntaxTests/iceRegressionTests/oversized_var.sol]: All the others are actually required to trigger this, though, I think.

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
Bugfix lien in the changelog? ;)

**Comment by chriseth** [comment]
it is not yet released.

**Comment by chriseth** [comment]
(was)

**Comment by ekpyron** [comment]
@axic It was only introduced after the last release.

File: test/libsolidity/syntaxTests/iceRegressionTests/oversized_var.sol [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    c d;
+    function e() public {
+        var d = d;
+    }
+}
+// ----
+// Warning 2519: (105-110): This declaration shadows an existing declaration.
+// Warning 3408: (66-69): Variable "d" covers a large part of storage and thus makes collisions likely. Either use mappings or dynamic arrays and allow their size to be increased only in small quantities per transaction.
+// Warning 2332: (105-110): Type "b.c" covers a large part of storage and thus makes collisions likely. Either use mappings or dynamic arrays and allow their size to be increased only in small quantities per transaction.
+// SyntaxError 1719: (105-114): Use of the "var" keyword is disallowed. Use explicit declaration `struct b.c storage pointer d = ...´ instead.
```
Comment (axic) [inline test/libsolidity/syntaxTests/iceRegressionTests/oversized_var.sol:15]: This message isn't that useful: `struct b.c storage pointer d` is hardly valid :)

Comment (a3d4) [reply test/libsolidity/syntaxTests/iceRegressionTests/oversized_var.sol:15]: Do we have a ticket for that?

Comment (chriseth) [reply test/libsolidity/syntaxTests/iceRegressionTests/oversized_var.sol:15]: This does net exist anymore in breaking.

**Pull request merged**